### ISSUE TITLE
feature: Skip some hooks when using `--all-files`

### ISSUE THREAD

**Issue by jalaziz** [pre-commit/pre-commit#2247]
I realize this may be a pretty far-fetched request, but I figured it couldn't hurt to start a discussion.

We'd like a configurable way to disable some hooks from running when using `pre-commit run --all-files`. However, the hook should still run when `--all-files` is not supplied or when it is explicitly specified as an arg.

## Use Case

We have a monorepo and some hooks (like terraform) have special dependencies that aren't always managed or installed by pre-commit. The files that the hooks run on are isolated to a specific folder. Most teams never interact with these files, but if they do, the hooks should trigger.

Occasionally, we like to run `pre-commit run --all-files` as hooks (or their dependencies) are upgraded to ensure everything is still passing. A recent example of why this may be needed was when we upgraded `black` and had files that were reformatted after the upgrade.

We periodically run into issues where engineers don't have the correct setup and those special hooks fail, resulting in confusion and blocked progress.

## Workarounds

I realize that we could use `SKIP` for this. The issue is that `SKIP` is opt-in and not configurable via `.pre-commit-config.yaml`. This means that the whole engineering team needs to set the variable and then unset it when they want to run the team-specific hooks. What we'd like to is sort of the opposite, have hooks that are skipped by default but can be opted-into.

## Risks

I realize this is a slippery slope and may not even be possible with the current architecture (I'm guilty of not taking a deeper look into the codebase yet). There's definitely a chance files get committed that should have been linted. We hope that by limiting the skipping to just the bare `pre-commit run --all-files` that this chance is sufficiently small that it is an acceptable trade off.

**Comment by asottile** [comment]
`SKIP` is designed for this, use that

**Comment by asottile** [comment]
(or set them to `stages: [manual]`)

**Comment by jalaziz** [comment]
@asottile The issue is `SKIP` doesn't actually work for the use case (not without having everyone set that in their environment). Neither does `stages: [manual]`.

`SKIP` is opt-in and `stages: [manual]` is nearly a full opt-out. There's no way to say "by default, don't run this when using `--all-files`" which is a bit more nuanced.

**Comment by asottile** [comment]
yeah, I don't intend to add built-in complexity for .01% use cases -- it's easy enough for you to write a small script or alias which does exactly what you want given the existing features

**Comment by jalaziz** [comment]
Totally understandable, and I would happily write a script (and maybe I'm missing something obvious), but:
1. How would the script know it's running in `all-files` mode vs staged mode?
2. How would the script call onto pre-existing hooks?

For (2), it seems like maybe we could have the hooks defined with `stages: [manual]` and then have a local script hook that calls `pre-commit run <hook>`.

**Comment by asottile** [comment]
`--all-files` is a conscious decision to run, so you'd only run the script in that case

**Comment by jalaziz** [comment]
> `--all-files` is a conscious decision to run, so you'd only run the script in that case

I think this is where the issue lies. We want to _not_ run something in that case and run that same something in all other cases.

Yes, it's a conscious decision, but that doesn't mean an entire engineering organization is aware that they should set `SKIP` in their local environment.

Ultimately, it seems like the workaround is a script that sets `SKIP` and trying to educate users to use that script instead of `pre-commit run --all-files`.

**Comment by asottile** [comment]
yeah I just mean you could have `./pre-commit-run-all-files` which is just `#!/usr/bin/env bash\nexec SKIP=... pre-commit run --all-files`

**Comment by jalaziz** [comment]
> yeah I just mean you could have `./pre-commit-run-all-files` which is just `#!/usr/bin/env bash\nexec SKIP=... pre-commit run --all-files`

Yup. Was hoping to keep it contained within the workflow of `pre-commit` (to avoid the inevitable person who doesn't read search internal docs first), but I completely understand it not being worth implementing.

**Comment by jalaziz** [comment]
Today we came across possibly a more appropriate use case for something related, but higher level. I'd be happy to start a new issue and contribute a PR if you agree.

Our UI team was using [`pretty-quick`](https://github.com/azz/pretty-quick) through `husky`. We've decided to consolidate on `pre-commit`. Instead of using pre-commit's prettier hook, we've decided to use a local hook that calls `yarn pretty-quick --staged` directly to avoid conflicting versions of `prettier` between `pre-commit` and our `package.json`. This is all fine and great. What has become apparent is `pre-commit run -a` won't do the _right_ thing now because [`--staged`](https://github.com/azz/pretty-quick#--staged-only-git) only works on staged files. What I believe would help solve this would be having a way to determine if the script is running in `all-files` mode so we could drop the `--staged` argument.

Looking at the source code, it seems like this should be doable by adding a [new environment variable](https://github.com/pre-commit/pre-commit/blob/master/pre_commit/commands/run.py#L363) that is passed to the hooks.
